test_plans/sub_group_mask.asciidoc

```diff
@@ -16,33 +16,106 @@ is selected on the CTS command line.
+
+
 
 == Tests
 
 All of the following tests have these initial steps performed in defferent .cpp files :
 
 * Create a `queue` from the tested device and call `queue::submit()`.
-* Submit a kernel via `handler::parallel_for()` with parameters `sycl::nd_range<1>(range<1>(128), range<1>(32))`
+* Submit a kernel via `handler::parallel_for()` with parameters `sycl::nd_range<1>(range<1>(32), range<1>(32))`
   and lambda finction with parameter `sycl::nd_item<1>`.
+* Lambda function should be decorated with attribute `reqd_sub_group_size(N)` with `N` values = `[8,16,32]`.
+  If device does not support `sub_group` size `N`, test for size `N` can be skipped.
+  `device.get_info<sycl::info::device::sub_group_sizes>()` should be used
+  to check supported `sub_group` sizes.
 * In lambda function use `nd_item::get_sub_group()` to get `sycl::sub_group` instance.
-* Use `sub_group_ballot` to get `sub_group_mask` instance with predicate suitable for test.
 * Use accessor to buffer for type `bool` to save results.
-* For regular test run check is performed only for leader work-item
+* For regular test run check is performed only for leader work-item.
 * For full conformance test run check is performed for all work-items.
 
 === Test for max_bits
 
 Check that member `size_t max_bits` exists.
 
-=== Tests for Member Functions
+=== Tests for constructors and operator= [[ctors_rev_2]]
+
+Features described in this paragraph are available from revision 2 of the
+extension.
+
+==== sub_group_mask() [[default_ctor]]
+
+* Construct `sub_group_mask` instance `s` `with sub_group_mask()` ctor.
+* Check that `s.none() == true`.
+
+==== sub_group_mask(unsigned long long val) [[ull_ctor]]
+
+* Construct `sub_group_mask` instance `s` with `sub_group_mask(unsigned long long val)` ctor
+  and `size_t val`. Where `val` is `item.get_local_linear_id()` and  `0b1010...`.
+* Check that for `N = 0...sub_group_mask.size() - 1` `s[id(N)]` == `((val >> N) & 1)`.
+
+==== template <typename T, std::size_t K> sub_group_mask(const sycl::marray<T, K>& val) [[marray_ctor]]
+
+For `marray<T, dim>` where `T` equals to integral types below:
+
+* `char`
+* `signed char`
+* `unsigned char`
+* `short int`
+* `unsigned short int`
+* `int`
+* `unsigned int`
+* `long int`
+* `unsigned long int`
+* `long long int`
+* `unsigned long long int`
+
+and `dim` is `2`, `5`, and `10`.
+
+* Construct `marray<T, dim>` instance `val` with default ctor.
+* Use initial variable `size_t value` with values `item.get_local_linear_id()` and `0b1010...`.
+* If `CHAR_BIT * sizeof(val) < CHAR_BIT * sizeof(value)`
+  for `N = CHAR_BIT * sizeof(val)` set first `N` bits of `val` to first `N` bits of `value`.
+* If `CHAR_BIT * sizeof(val) >= CHAR_BIT * sizeof(value)`
+  for `N = CHAR_BIT * sizeof(value)` set first `N` bits of `val` to first `N` bits of `value`.
+* Construct `sub_group_mask` instance `s` with `sub_group_mask(val)` ctor.
+* Check that for `N = 0...sub_group_mask.size() - 1` `s[id(N)]` == `((val >> N) & 1)`.
```

Comment:
This is OK too, but I think it would be clearer like this:

> Run the following test with all combinations of:
> * TYPE: `char`, `signed char`, ....
> * DIM: `2`, `5`, `10`
> * VALUE: `item.get_local_linear_id()`, `0b101010...`
> ```
> marray<TYPE, DIM> ma;
> size_t val = VALUE;
> if (sizeof(ma) < sizeof(val)) {
>   int N = CHAR_BIT * sizeof(ma);
>   /* set first N bits of "ma" to first N bits of "val" */
> } else {
>   int N = CHAR_BIT * sizeof(val);
>   /* set first N  bits of "ma" to first N bits of "val" */
> }
> sub_group_mask mask{ma};
> for (size_t i = 0;  i < mask.size();  i++) {
>   if (mask[id(i)] != ((val >> i) & 1)) {
>     /* test fails */
>   }
> }
> ```

Reply:
Fixed it. Thanks!